bin/lsif_indexer.dart

```diff
@@ -28,15 +28,34 @@
 // Licensed under the BSD-2 Clause License: https://github.com/astashov/crossdart/blob/master/LICENSE
 
 import 'dart:io';
+import 'package:path/path.dart';
+
 import 'package:lsif_indexer/analyzer.dart';
+import 'package:lsif_indexer/src/arguments.dart';
+import 'package:lsif_indexer/src/emitter.dart';
+import 'package:lsif_indexer/src/util/path_extensions.dart';
 
-/// Generate LSIF information for the directory listed in the first [argument], or
+/// Generate LSIF information for the directory provided from the [arguments], or
 /// the current directory if not specified.
 ///
-/// Currently prints to standard out.
+/// The format of the output is based on the [arugments].
 void main(List<String> arguments) async {
-  // TODO: Allow specifying an output file.
-  await Analyzer(
-    arguments.isEmpty ? Directory.current.absolute.path : arguments.first,
-  ).analyzePackage();
+  final config = LsifDartArgumentParser().parse(arguments);
+
+  // Exit early if the arguments were invalid
+  if (!config.isValid) return;
+
+  if (config.output != null) {
+    emitter = Emitter.fileOutput(config.output);
+  } else {
+    emitter = Emitter.standardOutput();
+  }
+
+  await emitter.use(() async {
+    await Analyzer(
```

Comment:
This is big enough, and kind of the main entrypoint, so it might be good to have it in a named function.

Reply:
Is this along the lines of what you were thinking?

```Dart
void main(List<String> arguments) async {
  final config = ArgumentParser().parse(arguments);

  // Exit early if the arguments were invalid
  if (!config.isValid) return;

  emitter = config.output == null
      ? Emitter.standardOutput()
      : Emitter.fileOutput(config.output);

  await _analyze(
    packageRoot: config.projectRoot?.absolute?.normalized,
    filesToAnalyze: config.rest.map(absolute).toList(),
  );
}

Future<void> _analyze({
  String packageRoot,
  List<String> filesToAnalyze,
}) async {
  await emitter.use(() async {
    await Analyzer(
      packageRoot: packageRoot ?? Directory.current.absolute.path,
      filesToAnalyze: filesToAnalyze,
    ).analyzePackage();
  });
}
```

Reply:
Actually I just meant lifting out the part within the `use`. so   
`  await emitter.use(_analyze);`
where _analyze is the whole body that was there before, as it can be harder to follow nested anonymous functions, especially with async.